panic in antiIdle while reconnecting in adverse network conditions

I had some network problems recently, and matterircd crashed while it was trying to reconnect:

```
time="2020-10-13T13:18:16+13:00" level=info msg="Found version 5.27.0.5.27.0.3bd4fc983a4aa622183c3bd1e77029e0.true" prefix=matterclient
time="2020-10-13T13:18:26+13:00" level=error msg="reconnect: login failed: https://chat.example.net/api/v4/users/me: model.client.connecting.app_error, Get "https://chat.example.net/api/v4/users/me": context deadline exceeded (Client.Timeout exceeded while awaiting headers), retrying in 10 seconds" prefix=matterclient
time="2020-10-13T13:18:27+13:00" level=error msg="ChannelView update for yryoj7ka6pr4ukbtojimq8ck8c failed: https://chat.example.net/api/v4/channels/members/ke9yfckxsbrkuyaa9hx4xd3use/view: model.client.connecting.app_error, Post "https://chat.example.net/api/v4/channels/members/ke9yfckxsbrkuyaa9hx4xd3use/view": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" prefix=matterclient
time="2020-10-13T13:18:36+13:00" level=info msg="reconnect: login" prefix=matterclient
time="2020-10-13T13:18:46+13:00" level=error msg="reconnect: login failed: "https://chat.example.net/api/v4/users/logout: model.client.connecting.app_error, Post \"https://chat.example.net/api/v4/users/logout\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)", retrying in 10 seconds" prefix=matterclient
time="2020-10-13T13:18:56+13:00" level=info msg="reconnect: login" prefix=matterclient
time="2020-10-13T13:19:06+13:00" level=error msg="reconnect: login failed: "https://chat.example.net/api/v4/users/logout: model.client.connecting.app_error, Post \"https://chat.example.net/api/v4/users/logout\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)", retrying in 10 seconds" prefix=matterclient
time="2020-10-13T13:19:16+13:00" level=info msg="reconnect: login" prefix=matterclient
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x809cbf]

goroutine 70 [running]:
github.com/42wim/matterircd/pkg/matterclient.(*Client).UpdateLastViewed(0xc0002c0000, 0xc0002d12a0, 0x1a, 0xb, 0xc000e67f30)
	/home/paul/go/src/github.com/42wim/matterircd/pkg/matterclient/channels.go:278 +0x11f
github.com/42wim/matterircd/bridge/mattermost.(*Mattermost).antiIdle(0xc0002be090, 0xc0002d12a0, 0x1a, 0xc0001f40c0)
	/home/paul/go/src/github.com/42wim/matterircd/bridge/mattermost/mattermost.go:188 +0x107
created by github.com/42wim/matterircd/bridge/mattermost.(*Mattermost).loginToMattermost
	/home/paul/go/src/github.com/42wim/matterircd/bridge/mattermost/mattermost.go:104 +0x4d9
```
Here's where the crash happens:

```go
func (m *Client) UpdateLastViewed(channelID string) error {
	m.logger.Debugf("posting lastview %#v", channelID)

	view := &model.ChannelView{ChannelId: channelID}

	for {
		_, resp := m.Client.ViewChannel(m.User.Id, view)	// <-- HERE

	// [...]
	}
}
```

I've added some code locally to check which of `m.Client` or `m.User` is nil and tried blackholing the server IPs but so far have been unable to reproduce the crash.

Possibly it would be safest to shut down `antiIdle` while reconnecting, but it's part of the top-level generic code, and the reconnect loop is handled in the bridge code, so this is not entirely trivial to do and might still be racy. So probably the first step is still to work out what is nil in the first place.

--- a/bridge/mattermost/mattermost.go
+++ b/bridge/mattermost/mattermost.go
@@ -51,6 +51,10 @@ func New(v *viper.Viper, cred bridge.Credentials, eventChan chan *bridge.Event, 
 		mc.SetLogLevel("debug")
 	}
 
+	if v.GetBool("trace") {
+		mc.SetLogLevel("trace")
+	}
+
 	m.mc = mc
 	m.connected = true
 
@@ -63,6 +67,8 @@ func (m *Mattermost) loginToMattermost(onWsConnect func()) (*matterclient.Client
 		mc.Credentials.NoTLS = true
 	}
 
+	// do anti idle on town-square, every installation should have this channel
+	mc.AntiIdle = !m.v.GetBool("mattermost.DisableAutoView")
 	mc.OnWsConnect = onWsConnect
 
 	if m.v.GetBool("debug") {
@@ -94,17 +100,6 @@ func (m *Mattermost) loginToMattermost(onWsConnect func()) (*matterclient.Client
 	m.quitChan = append(m.quitChan, quitChan)
 
 	go m.handleWsMessage(quitChan)
-
-	// do anti idle on town-square, every installation should have this channel
-	channels := m.mc.GetChannels()
-	for _, channel := range channels {
-		if channel.Name == "town-square" && !m.v.GetBool("mattermost.DisableAutoView") {
-			idleChan := make(chan struct{})
-			m.quitChan = append(m.quitChan, idleChan)
-			go m.antiIdle(channel.Id, idleChan)
-			continue
-		}
-	}
 
 	return mc, nil
 }

--- a/pkg/matterclient/matterclient.go
+++ b/pkg/matterclient/matterclient.go
@@ -63,6 +63,7 @@ type Client struct {
 	Users         map[string]*model.User
 	MessageChan   chan *Message
 	WsClient      *model.WebSocketClient
+	AntiIdle      bool
 	WsQuit        bool
 	WsConnected   bool
 	OnWsConnect   func()
@@ -160,6 +161,17 @@ func (m *Client) Login() error {
 	}
 
 	go m.checkConnection(ctx)
+
+	if m.AntiIdle {
+		channels := m.GetChannels()
+		for _, channel := range channels {
+			if channel.Name == "town-square" {
+				go m.antiIdle(ctx, channel.Id)
+
+				continue
+			}
+		}
+	}
 
 	return nil
 }
@@ -684,4 +696,23 @@ func (m *Client) HandleRatelimit(name string, resp *model.Response) error {
 	time.Sleep(time.Duration(waitTime) * time.Second)
 
 	return nil
+}
+
+func (m *Client) antiIdle(ctx context.Context, channelID string) {
+	m.logger.Debugf("starting antiIdle for %s", channelID)
+
+	ticker := time.NewTicker(time.Second * 60)
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.logger.Debugf("antiIlde: ctx.Done() triggered, exiting for %s", channelID)
+
+			return
+		case <-ticker.C:
+			m.logger.Tracef("antiIdle %s", channelID)
+
+			m.UpdateLastViewed(channelID)
+		}
+	}
 }

